### ISSUE TITLE
[JRuby] Padrino Thread-Safety Issue

### ISSUE THREAD

**Issue by sgonyea** [padrino/padrino-framework#863]
I've come across a thread-safety issue in Padrino, running on JRuby (1.6.7.2).  I don't know if this affects MRI or Rubinius.  I run into the issue using WEBrick, Puma, and Trinidad.  This issue seems to occur right after boot up.  Perhaps Padrino is lazily compiling routes, and should instead do so on boot-up?  That's just a guess; I have not dived deep enough into Padrino.

Anyway, if I boot up my app server and, once it's listening, I immediately run a script like the following:

```bash
curl "http://localhost:50018/foo/bar" &
curl "http://localhost:50018/foo/bar" &
curl "http://localhost:50018/foo/bar" &
curl "http://localhost:50018/foo/bar" &
curl "http://localhost:50018/foo/bar" &
```
Note that the '&' basically causes them to happen at nearly the same time (I know you probably know).

This blows up for me every time.  If I update the script to look like:

```bash
curl "http://localhost:50018/foo/bar"
curl "http://localhost:50018/foo/bar" &
curl "http://localhost:50018/foo/bar" &
curl "http://localhost:50018/foo/bar" &
curl "http://localhost:50018/foo/bar" &
```

Then the first curl will block the rest of the script, until the app server completes its request.  The rest of the script then executes in parallel, with no problems.

I was able to fix this thread-safety issue by wrapping `Padrino::Routing::ClassMethods#compiled_router` in a Mutex, such that it now becomes:
```ruby
      # Compiles the routes including deferred routes.
      @@compiled_router_mutex = Mutex.new
      def compiled_router
        @@compiled_router_mutex.synchronize {
          if deferred_routes.empty?
            return router
          else
            deferred_routes.each { |_, routes| routes.each { |(route, dest)| route.to(dest) } }
            @deferred_routes = nil
            return router
          end
        }
      end
```

I don't believe this is the correct solution, so I'm not submitting this as a pull-request.  I'd wager that this mutex should live somewhere deeper.

**Comment by dariocravero** [comment]
Are we good to remove [this patch](https://github.com/padrino/padrino-framework/blob/master/padrino-core/lib/padrino-core/application.rb#L29-L40) now that we're on the new router? @namusyaka @ujifgc

**Comment by namusyaka** [comment]
I'm going to confirm that in a few days.

**Comment by ujifgc** [comment]
I checked on `jruby 1.7.18`. Padrino `0.12.4` without the patch still fails, Padrino `0.13.0.rc1` without the patch works fine.

**Comment by dariocravero** [comment]
Happy days! :)

**Comment by namusyaka** [comment]
@ujifgc Great! Thank you!!

**Comment by mitchellhenke** [comment]
I'm trying to use jruby on padrino, but it doesn't look like there is a Padrino `0.13.0.rc1` branch or a release for it in rubygems?

**Comment by nesquena** [comment]
Not yet, I plan on putting together a 0.13.0.beta1 soon though.
